File: CTRW/SSA_Canard.py
"""
Python translation of the MATLAB script SSA_Canard.m.
Implements the Monte Carlo (MC) sampler, the stochastic simulation algorithm
(SSA) on a uniform grid, and the sparse linear system construction used to
combine SSA data with a PDE-based correction.

Dependencies: numpy, matplotlib, scipy (for sparse matrices).
"""

from __future__ import annotations

import math
import time
from typing import Tuple

import numpy as np
import matplotlib.pyplot as plt
from scipy import sparse
from scipy.sparse.linalg import lsqr


# ---------------------------- Vector field --------------------------------- #
def canard(x: np.ndarray) -> np.ndarray:
    """Vector field for the 2D ODE."""
    delta = 0.1
    a = 1 - delta / 8 - 3 * delta**2 / 32 - 173 * delta**3 / 1024 - 0.01
    y = np.zeros(2, dtype=float)
    y[0] = (x[0] + x[1] - x[0] ** 3 / 3.0) / delta
    y[1] = a - x[0]
    return y


# ---------------------------- Monte Carlo EM ------------------------------- #
def mc_2d(lowx: float, lowy: float, span: float, n: int, eps: float, sample: int, dt: float) -> np.ndarray:
    """Euler-Maruyama Monte Carlo approximation of the invariant density."""
    h = span / n
    data = np.zeros(n * n, dtype=float)
    counts = np.zeros((n, n), dtype=float)
    x_old = np.array([1.0, 1.0], dtype=float)
    sqrt_dt = math.sqrt(dt)

    for _ in range(sample):
        x_new = x_old + dt * canard(x_old) + eps * sqrt_dt * np.random.randn(2)
        xx, yy = x_new
        x_n = math.ceil((xx - lowx) / h)
        y_n = math.ceil((yy - lowy) / h)
        if 1 <= x_n <= n and 1 <= y_n <= n:
            counts[x_n - 1, y_n - 1] += 1
        x_old = x_new

    # flatten counts into data
    data[:] = counts.ravel(order="F")  # MATLAB is column-major; match it
    total = data.sum()
    if total > 0:
        data /= (h**2 * total)
    print(f"Sample Size = {sample}")
    return data


# ---------------------------- SSA on grid ---------------------------------- #
def ssa_2d(lowx: float, lowy: float, span: float, n: int, eps: float, sample: int) -> np.ndarray:
    """CTRW/SSA approximation of the invariant density on a uniform grid."""
    h = span / n
    data = np.zeros((n + 1) * (n + 1), dtype=float)
    counts = np.zeros((n + 1, n + 1), dtype=float)

    x = 1.0
    y = 1.0
    t = 0.0
    t_stop = 50000.0
    n_events = 0
    delta = 0.1
    a = 1 - delta / 8 - 3 * delta**2 / 32 - 173 * delta**3 / 1024 - 0.01
    sigma = eps
    m11 = sigma**2 / 2.0
    m22 = sigma**2 / 2.0
    c1 = m11 / h**2
    c2 = m22 / h**2

    while t < t_stop:
        mu_1 = (y - (x**3) / 3.0 + x) / delta
        mu_2 = a - x

        q1 = max(mu_1, 0.0) / h + c1
        q2 = -min(mu_1, 0.0) / h + c1
        q3 = max(mu_2, 0.0) / h + c2
        q4 = -min(mu_2, 0.0) / h + c2

        lam = q1 + q2 + q3 + q4
        r1, r2 = np.random.rand(2)
        tau = -math.log(r1) / lam

        # Mimic MATLAB's reversed cumulative selection
        cum = np.cumsum([q1, q2, q3, q4])
        mu_number = int(np.sum(r2 * lam <= cum))

        t += tau
        if mu_number == 4:
            x += h
        elif mu_number == 3:
            x -= h
        elif mu_number == 2:
            y += h
        elif mu_number == 1:
            y -= h

        x_n = round((x - lowx) / h)
        y_n = round((y - lowy) / h)
        if 1 <= x_n <= n + 1 and 1 <= y_n <= n + 1:
            counts[x_n - 1, y_n - 1] += 1

        n_events += 1
        if n_events == sample:
            break

    data[:] = counts.ravel(order="F")
    total = data.sum()
    if total > 0:
        data /= (h**2 * total)
    print(f"Sample Size = {n_events}")
    return data


# ---------------------------- Matrix assembly ------------------------------ #
def matrix_2d(lowx: float, lowy: float, span: float, n: int, eps0: float) -> Tuple[sparse.csr_matrix, np.ndarray]:
    """Sparse matrix and RHS construction mirroring the MATLAB Matrix_2D."""
    eps = eps0**2 / 2.0
    h = span / n
    f1 = lambda x, y: canard(np.array([x, y]))[0]
    f2 = lambda x, y: canard(np.array([x, y]))[1]

    b = np.zeros((n - 1) * (n - 1) + 1, dtype=float)
    b[-1] = 1 / h**2
    num_entries = 5 * (n - 1) * (n - 1) + (n + 1) * (n + 1)
    rows = np.zeros(num_entries, dtype=int)
    cols = np.zeros(num_entries, dtype=int)
    vals = np.zeros(num_entries, dtype=float)
    count = 0

    # Interior points: i,j in 1..n-1 (0-based)
    for i0 in range(1, n):
        for j0 in range(1, n):
            xx = i0 * h + lowx
            yy = j0 * h + lowy
            row = (i0 - 1) * (n - 1) + (j0 - 1)

            # +x neighbor
            rows[count] = row
            cols[count] = (i0 + 1) * (n + 1) + j0
            vals[count] = -f1(xx + h, yy) / (2 * h) + eps / (h**2)
            count += 1

            # -x neighbor
            rows[count] = row
            cols[count] = (i0 - 1) * (n + 1) + j0
            vals[count] = f1(xx - h, yy) / (2 * h) + eps / (h**2)
            count += 1

            # +y neighbor
            rows[count] = row
            cols[count] = i0 * (n + 1) + (j0 + 1)
            vals[count] = -f2(xx, yy + h) / (2 * h) + eps / (h**2)
            count += 1

            # -y neighbor
            rows[count] = row
            cols[count] = i0 * (n + 1) + (j0 - 1)
            vals[count] = f2(xx, yy - h) / (2 * h) + eps / (h**2)
            count += 1

            # center
            rows[count] = row
            cols[count] = i0 * (n + 1) + j0
            vals[count] = -4 * eps / (h**2)
            count += 1

    # Normalization row (all ones)
    for i0 in range(n + 1):
        for j0 in range(n + 1):
            rows[count] = (n - 1) * (n - 1)
            cols[count] = i0 * (n + 1) + j0
            vals[count] = 1.0
            count += 1

    A = sparse.coo_matrix(
        (vals[:count], (rows[:count], cols[:count])),
        shape=((n - 1) * (n - 1) + 1, (n + 1) * (n + 1)),
    ).tocsr()
    return A, b


# ---------------------------- Main driver ---------------------------------- #
def main() -> None:
    # Parameters (matching the MATLAB script)
    n = 600  # grid resolution
    eps = 0.3  # noise strength
    lowx = -3.0
    lowy = -3.0
    span = 6.0
    sample = 10_000_000  # This is very large; reduce for quicker tests
    dt = 0.001

    # SSA sampler (Monte Carlo section is commented out in the MATLAB code)
    t1 = time.perf_counter()
    data2 = ssa_2d(lowx, lowy, span, n, eps, sample)
    t2 = time.perf_counter()
    print("SSA_2D data generated")
    print(f"Elapsed time: {t2 - t1:.2f} seconds")

    # Build matrix and solve the least-norm problem
    A, b = matrix_2d(lowx, lowy, span, n, eps)
    print("matrix built")
    b = b - A @ data2
    # least-norm solution for the sparse system
    x = lsqr(A, b, atol=1e-10, btol=1e-10, iter_lim=10_000)[0]
    y = x + data2

    V = y.reshape((n + 1, n + 1), order="F")
    W = data2.reshape((n + 1, n + 1), order="F")

    fig = plt.figure(figsize=(10, 8))
    ax1 = fig.add_subplot(2, 1, 1, projection="3d")
    ax1.plot_surface(
        *np.meshgrid(np.arange(n + 1), np.arange(n + 1), indexing="ij"),
        W,
        cmap="viridis",
    )
    ax1.set_title("Monte Carlo estimate W")

    ax2 = fig.add_subplot(2, 1, 2, projection="3d")
    ax2.plot_surface(
        *np.meshgrid(np.arange(n + 1), np.arange(n + 1), indexing="ij"),
        V,
        cmap="viridis",
    )
    ax2.set_title("Corrected solution V")
    plt.tight_layout()
    plt.show()


if __name__ == "__main__":
    main()

File: 研究论文/CTRW/师姐程序/SSA_Canard.m
N = 600;%NxN Grid 
eps = 0.3;%Strength of noise % 0.5 original 
lowx = -3.0;%left end of numerical domain
lowy = -3.0;%bottom end of numerical domain
Sp = 6.0;%length (and height) of the numerical domain
Sample = 10000000;%Sample size of Monte Carlo simulator
dt = 0.001;%time step of numerical ODE scheme in MC simulator
clf
%{
t1=clock;
data1 = MC_2D(lowx, lowy, Sp, N, eps, Sample, dt);%Generate approximate data
t2=clock;
disp('MC_2D data generated');
computer_running_time=etime(t2,t1)
%}
t1=clock;
data2 = SSA_2D(lowx, lowy, Sp, N, eps, Sample);
t2=clock;
disp('SSA_2D data generated');
computer_running_time=etime(t2,t1)


[A, b] = Matrix_2D(lowx, lowy, Sp, N, eps);%Build matrix
disp('matrix built');
b = b - A*data2;
x = lsqminnorm(A,b);%Find the least norm solution
y = x + data2;
V = zeros(N+1,N+1);%Numerical solution
W = zeros(N+1,N+1);%A rough Solution obtained by Monte Carlo
for i = 1:N+1
    for j = 1:N+1
        V(i,j) = V(i,j) + y((i-1)*(N+1) + j);
        W(i,j) = W(i,j) + data2((i-1)*(N+1)+j);
    end
end
subplot(2,1,1)
mesh(W)
subplot(2,1,2)
mesh(V)

function y = canard(x)%ODE
delta = 0.1;
a = 1 - delta/8 - 3*delta^2/32 - 173*delta^3/1024 - 0.01;
y = zeros(2,1);
y(1) = (x(1) + x(2) - x(1)^3/3)/delta;
y(2) = a - x(1);
end

function data1 = MC_2D(lowx, lowy, Sp, N, eps, Sample, dt)
h = Sp/N;
data = zeros(N*N,1);
P = zeros(N,N);
x_old = [1 1]';

%xplot=1;% For plotting
%yplot=1;% For plotting
for i = 1:Sample
    x_new = x_old + dt*canard(x_old) + eps*sqrt(dt)*randn(2,1);
    xx = x_new(1);
    yy = x_new(2);
    x_n = ceil((xx - lowx)/h);
    y_n = ceil((yy - lowy)/h);
    if x_n >= 1 && x_n <=N && y_n >= 1 && y_n <= N
        P(x_n, y_n) = P(x_n,y_n) + 1;%Count number of sample points in each bin
    end
    x_old = x_new;   
    %xplot(i+1) = x_new(1);% For plotting
    %yplot(i+1) = x_new(2);% For plotting
end
for i = 1:N
    for j = 1:N
        data((i-1)*N + j) = P(i,j);
    end
end
data1 = data/(h^2*sum(data));%Normalization
fprintf('Sample Size = %d\n',Sample);
%{
subplot(2,2,1)
plot(xplot,yplot)
%}
subplot(2,1,1)
P=P/(h^2*sum(data));
mesh(P)

end

function data2 = SSA_2D(lowx, lowy, Sp, N, eps, Sample)
%%[dX,dY]=[1/epsilon*(Y-1/3*X^3+X),(a-X)]dt+sigma*dW, 
% dY=mu(Y)dt +sqrt(2)sigma(Y)dW in (2.1.2) in Vanden-Eijnden's paper
% mu(x,y)=[10*(Y-1/3*X^3+X),(0.9964-X)],and
% M_{1,1}=sigma^2/2, M_{1,2}=0, M_{2,1}=0, M_{2,2}=sigma^2/2, 
%--------------------------------------------------------------------------

h = Sp/N;%delta_x=delta_x_plus=delta_x_minus=h;%delta_y=delta_y_plus=delta_y_minus=h;
data = zeros((N+1)*(N+1),1);
P = zeros(N+1,N+1);
x=1;y=1;%initial value
t=0;%Time variable
%{
tplot=0;
xplot=x;
yplot=y;% For plotting
%}
t_stop=50000;    
n=0; %Reaction counter.
delta = 0.1;
a = 1 - delta/8 - 3*delta^2/32 - 173*delta^3/1024 - 0.01;
sigma=eps;
M_11=sigma^2/2; M_22=sigma^2/2;
%% STEP 1
C1=(M_11/h^2); C2=(M_22/h^2);
while t<t_stop
    mu_1=(y-1/3*x^3+x)/delta;%Q_u
    mu_2=a-x;%Q_u
    q1=max(mu_1,0)/h+C1;%Q_u
    q2=-min(mu_1,0)/h+C1;%Q_u
    q3=max(mu_2,0)/h+C2;%Q_u
    q4=-min(mu_2,0)/h+C2;%Q_u
    %{
    mu_tilde_1=20/sigma^2*(y-1/3*x^3+x);%Q_c
    mu_tilde_2=2/sigma^2*(a-x);%Q_c
    q1=exp(1/2*mu_tilde_1*h)*C1; % X_{i,j} -> X_{i+1,j)%Q_c
    q2=exp(-1/2*mu_tilde_1*h)*C1; % X_{i,j} -> X_{i-1,j)%Q_c
    q3=exp(1/2*mu_tilde_2*h)*C2; % X_{i,j} -> X_{i,j+1)%Q_c
    q4=exp(-1/2*mu_tilde_2*h)*C2; % X_{i,j} -> X_{i,j-1)%Q_c
    %}
    lambda=q1+q2+q3+q4; % four channels
    %% STEP 2
    r = rand(2,1);
    tau = -log(r(1))/lambda;% 
    mu_number = sum(r(2)*lambda<=cumsum([q1,q2,q3,q4]));
    %% STEP 3
    t = t + tau;
    % Adjust population levels.
    switch mu_number
        case 4
            x = x + h; % X_{i,j} -> X_{i+1,j)
        case 3
            x = x - h; % X_{i,j} -> X_{i-1,j)
        case 2
            y = y + h; % X_{i,j} -> X_{i,j+1)
        case 1
            y = y - h; % X_{i,j} -> X_{i,j-1)
    end
    x_n = round((x - lowx)/h);
    y_n = round((y - lowy)/h);
    if x_n >= 1 && x_n <=N+1 && y_n >= 1 && y_n <= N+1
        P(x_n, y_n) = P(x_n,y_n) + 1;%Count number of sample points in each bin
    end
    n = n + 1;
    if n==Sample
        break
    end
    %{
    tplot(n+1) = t;
    xplot(n+1) = x;
    yplot(n+1) = y;
    %}
end
    for i = 1:N+1
    for j = 1:N+1
        data((i-1)*(N+1) + j) = P(i,j);
    end
    end
    data2 = data/(h^2*sum(data));%Normalization
    fprintf('Sample Size = %d\n',n);
    %{
    subplot(2,2,3)
    plot(xplot,yplot)
    %}
    P=P/(h^2*sum(data));
    subplot(2,1,2)
    mesh(P)
    %plot3(tplot,xplot,yplot)
    
end

function [A,b]= Matrix_2D(lowx, lowy, Sp, N, eps0)
eps = eps0^2/2;%Coefficient of the Laplacian
h = Sp/N;%grid size
subindex = @(vec,index) vec(index);
f1 = @(x,y) subindex(canard([x y]'),1);%x component of the vector field
f2 = @(x,y) subindex(canard([x y]'),2);%y component of the vector field
%{
b = zeros((N-2)*(N-2) + 1,1);
b((N-2)*(N-2) + 1) = 1/h^2;
Num = 5*(N-2)*(N-2) + N*N;
%}
b = zeros((N-1)*(N-1) + 1,1);
b((N-1)*(N-1) + 1) = 1/h^2;
Num = 5*(N-1)*(N-1) + (N+1)*(N+1);
index1 = zeros(1,Num);
index2 = zeros(1,Num);
value = zeros(1,Num);
count = 1;
%{
for i = 2:N-1
    for j = 2:N-1
        xx = (i-1)*h + lowx ;%Important: A grid point is in the middle of a Monte Carlo bin
        yy = (j-1)*h + lowy ;
        index1(count) = (i-2)*(N-2) + j - 1;
        index2(count) = (i-1)*N + N + j;
        value(count) = -f1(xx+h,yy)/(2*h) + eps/(h^2);
        count = count +1;
        

        index1(count) = (i-2)*(N-2) + j - 1;
        index2(count) = (i-1)*N - N +j;
        value(count) = f1(xx-h,yy)/(2*h) + eps/(h^2);
        count = count +1;


        index1(count) = (i-2)*(N-2) + j - 1;
        index2(count) = (i-1)*N+j+1;
        value(count) = -f2(xx,yy+h)/(2*h) + eps/(h^2);
        count = count +1;

        
        index1(count) = (i-2)*(N-2) + j - 1;
        index2(count) = (i-1)*N + j -1;
        value(count) = f2(xx,yy-h)/(2*h) + eps/(h^2);
        count = count +1;

        
        index1(count) = (i-2)*(N-2) + j - 1;
        index2(count) = (i-1)*N+j;
        value(count) = -4*eps/(h^2);
        count = count +1;
    end
end
%}
for i = 2:N
    for j = 2:N
        xx = (i-1)*h + lowx ;%Important: A grid point is in the middle of a Monte Carlo bin
        yy = (j-1)*h + lowy ;
        index1(count) = (i-2)*(N-1) + j - 1;
        index2(count) = (i-1)*(N+1) + (N+1) + j;
        value(count) = -f1(xx+h,yy)/(2*h) + eps/(h^2);
        count = count +1;
        
        index1(count) = (i-2)*(N-1) + j - 1;
        index2(count) = (i-1)*(N+1) - (N+1) +j;
        value(count) = f1(xx-h,yy)/(2*h) + eps/(h^2);
        count = count +1;

        index1(count) = (i-2)*(N-1) + j - 1;
        index2(count) = (i-1)*(N+1)+j+1;
        value(count) = -f2(xx,yy+h)/(2*h) + eps/(h^2);
        count = count +1;

        index1(count) = (i-2)*(N-1) + j - 1;
        index2(count) = (i-1)*(N+1) + j -1;
        value(count) = f2(xx,yy-h)/(2*h) + eps/(h^2);
        count = count +1;
     
        index1(count) = (i-2)*(N-1) + j - 1;
        index2(count) = (i-1)*(N+1)+j;
        value(count) = -4*eps/(h^2);
        count = count +1;
    end
end
%{
for i = 1:N
    for j = 1:N
        index1(count) = (N-2)*(N-2) + 1;
        index2(count) = (i-1)*N + j;
        value(count) = 1;
        count = count + 1;
    end
end
%}
for i = 1:N+1
    for j = 1:N+1
        index1(count) = (N-1)*(N-1) + 1;
        index2(count) = (i-1)*(N+1) + j;
        value(count) = 1;
        count = count + 1;
    end
end
%A = sparse(index1,index2,value,(N-2)*(N-2) + 1,N*N);%Use sparse matrix
A = sparse(index1,index2,value,(N-1)*(N-1) + 1,(N+1)*(N+1));%Use sparse matrix
end

